storage/src/adb/any/fixed/mod.rs

```diff
@@ -256,3 +262,187 @@ where
+    /// the operation if it was active.
+    pub(crate) async fn move_op_if_active(
+        &mut self,
+        op: O,
+        old_loc: Location,
+    ) -> Result<Option<Location>, Error> {
+        // If the translated key is not in the snapshot, get a cursor to look for the key.
+        let Some(key) = op.key() else {
+            return Ok(None); // operations without keys cannot be active
+        };
+        let Some(mut cursor) = self.snapshot.get_mut(key) else {
+            return Ok(None);
+        };
+
+        // Find the snapshot entry corresponding to the operation.
+        if cursor.find(|&loc| *loc == old_loc) {
+            // Update the operation's snapshot location to point to tip.
+            let tip_loc = Location::new_unchecked(self.log.size().await?);
+            cursor.update(tip_loc);
+            drop(cursor);
+
+            // Apply the operation at tip.
+            self.apply_op(op).await?;
+            return Ok(Some(old_loc));
+        }
+
+        // The operation is not active, so this is a no-op.
+        Ok(None)
+    }
+
+    /// Raise the inactivity floor by taking one _step_, which involves searching for the first
+    /// active operation above the inactivity floor, moving it to tip, and then setting the
+    /// inactivity floor to the location following the moved operation. This method is therefore
+    /// guaranteed to raise the floor by at least one. Returns the new inactivity floor location.
+    ///
+    /// # Panics
+    ///
+    /// Expects there is at least one active operation above the inactivity floor, and panics otherwise.
+    async fn raise_floor(&mut self, mut inactivity_floor_loc: Location) -> Result<Location, Error>
+    where
+        E: Storage + Clock + Metrics,
+        I: Index<Value = Location>,
+        H: CHasher,
+        O: FixedOperation,
+    {
+        // Search for the first active operation above the inactivity floor and move it to tip.
+        //
+        // TODO(https://github.com/commonwarexyz/monorepo/issues/1829): optimize this w/ a bitmap.
+        loop {
+            let tip_loc = Location::new_unchecked(self.log.size().await?);
+            if tip_loc <= *inactivity_floor_loc {
+                panic!("no active operations above the inactivity floor");
+            }
+            let old_loc = inactivity_floor_loc;
+            inactivity_floor_loc += 1;
+            let op = self.log.read(*old_loc).await?;
+            if self.move_op_if_active(op, old_loc).await?.is_some() {
+                return Ok(inactivity_floor_loc);
+            }
+        }
+    }
+
+    /// Same as `raise_floor` but uses the status bitmap to more efficiently find the first active
+    /// operation above the inactivity floor.
+    ///
+    /// # Panics
+    ///
+    /// Panics if there is not at least one active operation above the inactivity floor.
+    pub(crate) async fn raise_floor_with_bitmap<const N: usize>(
```

Comment:
`raise_floor_with_bitmap`, `sync_and_process_updates`, and `sync` all take ownership of `self` whereas the other methods take in `&mut self`. This doesn't affect correctness but it might be cleaner to make them all parameterized on one or the other?